daemon/api_connections.go

```diff
@@ -0,0 +1,229 @@
+
+import (
+	"net/http"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/ifacestate"
+)
+
+var connectionsCmd = &Command{
+	Path:   "/v2/connections",
+	UserOK: true,
+	GET:    getConnections,
+}
+
+type collectFilter struct {
+	snapName  string
+	ifaceName string
+	connected bool
+}
+
+func (c *collectFilter) plugMatches(plug *interfaces.PlugRef, connectedSlots []interfaces.SlotRef) bool {
+	for _, slot := range connectedSlots {
+		if c.slotMatches(&slot, nil) {
+			return true
+		}
+	}
+	if c.snapName != "" && plug.Snap != c.snapName {
+		return false
+	}
+	return true
+}
+
+func (c *collectFilter) slotMatches(slot *interfaces.SlotRef, connectedPlugs []interfaces.PlugRef) bool {
+	for _, plug := range connectedPlugs {
+		if c.plugMatches(&plug, nil) {
+			return true
+		}
+	}
+	if c.snapName != "" && slot.Snap != c.snapName {
+		return false
+	}
+	return true
+}
+
+func (c *collectFilter) ifaceMatches(ifaceName string) bool {
+	if c.ifaceName != "" && c.ifaceName != ifaceName {
+		return false
+	}
+	return true
+}
+
+type bySlotRef []interfaces.SlotRef
+
+func (b bySlotRef) Len() int      { return len(b) }
+func (b bySlotRef) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+func (b bySlotRef) Less(i, j int) bool {
+	return b[i].SortsBefore(b[j])
+}
+
+type byPlugRef []interfaces.PlugRef
+
+func (b byPlugRef) Len() int      { return len(b) }
+func (b byPlugRef) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+func (b byPlugRef) Less(i, j int) bool {
+	return b[i].SortsBefore(b[j])
+}
+
+func collectConnections(ifaceMgr *ifacestate.InterfaceManager, filter collectFilter) (*connectionsJSON, error) {
+	repo := ifaceMgr.Repository()
+	ifaces := repo.Interfaces()
+
+	var connsjson connectionsJSON
+	var connStates map[string]ifacestate.ConnectionState
+	plugConns := map[string][]interfaces.SlotRef{}
+	slotConns := map[string][]interfaces.PlugRef{}
+
+	var err error
+	connStates, err = ifaceMgr.ConnectionStates()
+	if err != nil {
+		return nil, err
+	}
+
+	connsjson.Established = make([]connectionJSON, 0, len(connStates))
+	for crefStr, cstate := range connStates {
+		cref, err := interfaces.ParseConnRef(crefStr)
+		if err != nil {
+			return nil, err
+		}
+		if !filter.plugMatches(&cref.PlugRef, nil) && !filter.slotMatches(&cref.SlotRef, nil) {
+			continue
+		}
+		if !filter.ifaceMatches(cstate.Interface) {
+			continue
+		}
+		plugRef := interfaces.PlugRef{Snap: cref.PlugRef.Snap, Name: cref.PlugRef.Name}
+		slotRef := interfaces.SlotRef{Snap: cref.SlotRef.Snap, Name: cref.SlotRef.Name}
+		plugID := plugRef.String()
+		slotID := slotRef.String()
+
+		cj := connectionJSON{
+			Slot:      slotRef,
+			Plug:      plugRef,
+			Manual:    cstate.Auto == false,
+			Gadget:    cstate.ByGadget,
+			Interface: cstate.Interface,
+		}
+		if cstate.Undesired {
+			// explicitly disconnected are always manual
+			cj.Manual = true
+			connsjson.Undesired = append(connsjson.Undesired, cj)
```

Comment:
should we skip filling this if connected == true?

Reply:
These should be skipped now.